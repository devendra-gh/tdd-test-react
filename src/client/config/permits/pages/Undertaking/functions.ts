import { IVariables } from '@tamm/app-composer';
import PermitConfigs from 'client/config/permits/permitConfigs';
import baseUrl from 'client/utils/baseUrl';
import { PATH_APPLICATION_DETAILS } from '../../utils/constants/pageRoutes';

const permitConfigs: { [key: string]: any } = PermitConfigs;

interface IServiceConfig {
  name: string;
  category: string;
  requiresUndertakingApproval: boolean;
}

const onInit = (props: IVariables) => {
  const { serviceName } = props.match.params;
  const { serviceType, permitType } = props;
  if (serviceName === '' && serviceType === '') {
    window.location.href = baseUrl;
  } else {
    const serviceConfig =
      serviceName === ''
        ? permitConfigs[serviceType]
        : permitConfigs[serviceName];
    if (serviceConfig) {
      const {
        category,
        requiresUndertakingApproval,
        name: serviceNameConfig,
      } = serviceConfig;
      if (permitType !== category) props.actions.permitType.update(category);
      if (serviceType !== serviceNameConfig)
        props.actions.serviceType.update(serviceNameConfig);
      if (!requiresUndertakingApproval)
        props.history.push(`/${serviceNameConfig}${PATH_APPLICATION_DETAILS}`);
    } else window.location.href = baseUrl;
  }
};

const handleToggleCheckbox = (props: IVariables) => {
  const { serviceType, permitInfo } = props;
  props.actions.permitInfo.update({
    ...permitInfo,
    [serviceType]: {
      ...permitInfo[serviceType],
      undertaking: {
        ...permitInfo[serviceType].undertaking,
        isApproved: !permitInfo[serviceType].undertaking.isApproved,
      },
    },
  });
};
const onCancelClick = (e: Event, props: IVariables) => {
  const { serviceType, permitInfo } = props;
  props.actions.permitInfo.update({
    ...permitInfo,
    [serviceType]: {
      ...permitInfo[serviceType],
      undertaking: {
        ...permitInfo[serviceType].undertaking,
        isApproved: false,
      },
    },
  });
};
const onSubmit = (e: Event, props: IVariables) => {
  const { serviceType, permitInfo } = props;
  const { requiresUndertakingApproval } = permitConfigs[serviceType];
  const { undertaking } = permitInfo[serviceType];
  props.actions.permitInfo.update({
    ...permitInfo,
    [serviceType]: {
      ...permitInfo[serviceType],
      undertaking: {
        ...permitInfo[serviceType].undertaking,
        showError: !undertaking.isApproved,
      },
    },
  });

  if (requiresUndertakingApproval && undertaking.isApproved) {
    props.history.push(`/${serviceType}${PATH_APPLICATION_DETAILS}`);
  }
};
export default { onInit, handleToggleCheckbox, onSubmit, onCancelClick };
