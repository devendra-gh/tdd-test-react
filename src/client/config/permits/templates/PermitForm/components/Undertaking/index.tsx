import React from 'react';
import { IVariables } from '@tamm/app-composer';
import Form from '@tamm/ui-lib-v2-form';
import Checkbox from '@tamm/ui-lib-v2-checkbox';
import './undertaking.less';

/**
 * TermsConditionsList component
 * @param {Object} props
 * @returns {JSX}
 */
function Undertaking(props: IVariables) {
  const {
    inputGroups,
    handleChange,
    values,
    i18n,
    startShowingErrors,
    name: stateKey,
  } = props;
  const { [stateKey]: termsConditions } = values;

  const handleToggleCheckbox = (fieldName: string) => {
    return () => {
      handleChange({
        [stateKey]: {
          ...termsConditions,
          [fieldName]: !termsConditions[fieldName],
        },
      });
    };
  };

  return (
    inputGroups && (
      <Form name={props.name} id={props.name}>
        {inputGroups.map((formGroups: IVariables) => (
          <Form.Fieldset key={`${formGroups.groupKey}.${formGroups.stateKey}`}>
            {formGroups.fields.map((fileConfig: IVariables) => {
              const {
                name,
                label,
                validationConfig,
                ...restProps
              } = fileConfig;

              return (
                <React.Fragment key={`${name}.${formGroups.groupKey}`}>
                  <Checkbox
                    {...restProps}
                    checked={termsConditions[name]}
                    label={label}
                    name={name}
                    onChange={handleToggleCheckbox(name)}
                    validateStatus={
                      startShowingErrors &&
                      validationConfig &&
                      !termsConditions[name]
                        ? 'error'
                        : ''
                    }
                  />
                </React.Fragment>
              );
            })}
            <div>
              {formGroups.conditions && formGroups.conditions.length > 0 && (
                <ul className="PermitForm__form-tac-ui">
                  {formGroups.conditions.map(
                    (condition: string, key: number) => (
                      <li key={String(key)}>{i18n(condition)}</li>
                    ),
                  )}
                </ul>
              )}
              {formGroups.title && (
                <p className="termsTitle marginT20">{i18n(formGroups.title)}</p>
              )}
            </div>
          </Form.Fieldset>
        ))}
      </Form>
    )
  );
}

export default Undertaking;
