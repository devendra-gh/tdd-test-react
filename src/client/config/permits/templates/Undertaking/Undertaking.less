@import '~@tamm/ui-lib-v2-styles/colors.less';

.ui-lib-locale-provider_rtl .ui-lib-link {
  margin-right: 2rem;
}

ul.PermitForm__form-tac-ui {
  margin-bottom: 0;
  color: @ui-lib-body-text-color;
  line-height: 1.6;
  font-size: 1.5;
  padding-left: 0 !important;
}

.ui-lib-viewport ul.PermitForm__form-tac-ui li {
  display: flex;
  flex-direction: row;
  font-size: 16px;
}

.ui-lib-viewport ul.PermitForm__form-tac-ui li::before {
  // font-size: 1.5rem;
  padding: 0 1em;
  padding-top: 1.3rem;
  top: 0;
}

.ui-lib-locale-provider .ui-lib-link {
  margin-left: 2rem;
}
