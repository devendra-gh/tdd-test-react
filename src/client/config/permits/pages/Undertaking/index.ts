import functions from './functions';
import { PATH_UNDERTAKING } from '../../utils/constants/pageRoutes';
import { REQUIRES_CUSTOM_LOGIN } from '../../constants';

const undertaking = [
  {
    path: `/:serviceName${PATH_UNDERTAKING}`, // path for router
    uniqueId: 'permits-undertaking', // uniqueId for caching and other purposes
    template: 'undertaking', // template name, must be located in index of folder template/index
    title: 'pageTitle.permitApplication', // title of the page, later it will be read from CMS
    init: functions.onInit,
    props: {
      handleToggleCheckbox: functions.handleToggleCheckbox,
      onSubmit: functions.onSubmit,
      onCancelClick: functions.onCancelClick,
      startShowingErrors: false,
    },
    state: {
      mapState: [
        'loggedIn',
        'businessKey',
        'serviceType',
        'permitType',
        'permitInfo',
        'startShowingErrors',
      ],
      mapDispatch: [
        'instanceId',
        'businessKey',
        'serviceType',
        'permitType',
        'permitInfo',
        'startShowingErrors',
      ],
    },
    requires: [REQUIRES_CUSTOM_LOGIN],
  },
];

export default undertaking;
