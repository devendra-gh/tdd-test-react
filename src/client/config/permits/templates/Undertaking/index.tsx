import React from 'react';
import Button from '@tamm/ui-lib-v2-button';
import Container from 'client/containers';
import { withTemplateHooks, IVariables } from '@tamm/app-composer';
import routePropTypes from '@tamm/app-composer/client/propTypes/route';
import ServiceTemplate from '@tamm/ui-lib-v2-service-template';
import Form from '@tamm/ui-lib-v2-form';
import Checkbox from '@tamm/ui-lib-v2-checkbox';
import permitConfigs from 'client/config/permits/permitConfigs';
import Link from '@tamm/ui-lib-v2-link';
import { PATH_LANDING } from '../../utils/constants/pageRoutes';
import './Undertaking.less';

/**
 * Undertaking template
 * @param       {Object} props
 * @returns     {JSX}
 */
const Undertaking = (props: IVariables) => {
  const PermitConfigs: { [key: string]: any } = permitConfigs;
  const {
    i18n,
    serviceType,
    onSubmit,
    handleToggleCheckbox,
    permitInfo,
    onCancelClick,
  } = props;
  const serviceConfig = PermitConfigs[serviceType];
  const { title, subTitle, conditions } = serviceConfig.undertakings;
  const { isApproved, showError } = permitInfo[serviceType].undertaking;
  const name = 'isApproved';
  return (
    <Container>
      <Form name={props.name} id={props.name}>
        <Form.Fieldset>
          <ServiceTemplate title={i18n(title)} description={i18n(subTitle)} />
          <div>
            {conditions && conditions.length > 0 && (
              <ul className="PermitForm__form-tac-ui">
                {conditions.map((condition: string, key: number) => (
                  <li key={String(key)}>{i18n(condition)}</li>
                ))}
              </ul>
            )}
          </div>
          <Checkbox
            checked={isApproved}
            label={i18n('undertaking.agree')}
            onChange={e => {
              handleToggleCheckbox(props);
            }}
            name={name}
            validateStatus={showError && !isApproved ? 'error' : ''}
          />
          <div>
            <Button
              aria-label={i18n('continue')}
              label={i18n('continue')}
              onClick={e => {
                onSubmit(e, props);
              }}
            />
            <Link
              aria-label={i18n('cancel')}
              className="PermitForm__button-cancel"
              disabled={false}
              href={`/${serviceType}${PATH_LANDING}`}
              label={i18n('cancel')}
              tammHref="/"
              onClick={e => {
                onCancelClick(e, props);
              }}
              type={undefined}
            />
          </div>
        </Form.Fieldset>
      </Form>
    </Container>
  );
};

Undertaking.propTypes = {
  ...routePropTypes,
};

export default withTemplateHooks(Undertaking);
