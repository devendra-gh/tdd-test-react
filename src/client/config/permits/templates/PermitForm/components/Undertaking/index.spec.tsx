import React from 'react';
import { shallow } from 'enzyme';
import Undertaking from './index';
// import { checkValidationField } from 'client/config/utils/checkValidation';

// jest.mock('client/config/utils/checkValidation', () => ({
//   checkValidationField: {
//     REQUIRED: 'REQUIRED'
//   }
// }));

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Permit Type form', () => {
  let props: any;

  beforeEach(() => {
    props = {
      i18n: jest.fn(),
      handleToggleCheckbox: jest.fn(),
      values: {
        undertakings: {
          approved: true,
        },
      },
      inputGroups: [
        {
          name: 'test1',
          title: 'test',
          groupKey: 'Gkey',
          stateKey: 'Skey',
          conditions: ['term1'],
          fields: [
            {
              validationConfig: {
                type: 'REQUIRED',
              },
              elementType: 'checkBox',
              name: 'approved',
              label: 'fieldLalel',
              // onChange: () => jest.fn(),
            },
          ],
        },
      ],
      handleChange: jest.fn(),
      startShowingErrors: true,
      name: 'undertakings',
    };
  });

  it('should successfully render when serviceType is provided', () => {
    const component = shallow(<Undertaking {...props} />);
    expect(component).toMatchSnapshot();
  });
  // eslint-disable-next-line no-console
  console.log = () => {};
  console.info = () => {};
  console.warn = () => {};
  console.error = () => {};

  // describe('checkbox', () => {
  //   it('should handle checkbox change event with termsConditions.approved as false', () => {
  //     props.values.undertakings.approved = false;
  //     const component = shallow(<Undertaking {...props} />);

  //     const Checkbox = component.find('Memo(Checkbox)');
  //     const allProps: any = Checkbox.props();
  //     const { onChange } = allProps;
  //     onChange('approved');
  //     expect(props.handleToggleCheckbox).toBeTruthy();
  //   });
  // });
});
