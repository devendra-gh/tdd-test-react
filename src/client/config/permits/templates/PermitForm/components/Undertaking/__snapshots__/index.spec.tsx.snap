// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permit Type form should successfully render when serviceType is provided 1`] = `
<LocalizedComponent(Form)
  backButton={null}
  cancelLink={null}
  help=""
  i18n={[Function]}
  id="undertakings"
  name="undertakings"
  submitButton={null}
  validateStatus={null}
>
  <Form.Fieldset
    buttonGroup={false}
    gapSize="default"
    key="Gkey.Skey"
    title=""
    twoColumns={false}
  >
    <Checkbox
      checked={true}
      elementType="checkBox"
      label="fieldLalel"
      name="approved"
      onChange={[Function]}
      validateStatus=""
    />
    <div>
      <ul
        className="PermitForm__form-tac-ui"
      >
        <li
          key="0"
        />
      </ul>
      <p
        className="termsTitle marginT20"
      />
    </div>
  </Form.Fieldset>
</LocalizedComponent(Form)>
`;
